File: frontend/types/rules.ts
/**
 * 规则管理类型定义
 * 与后端 api/rules.py 对齐
 */

// ===== 违禁词 =====

export interface ForbiddenWordCreate {
  word: string
  category: string
  severity: string
}

export interface ForbiddenWordResponse {
  id: string
  word: string
  category: string
  severity: string
}

export interface ForbiddenWordListResponse {
  items: ForbiddenWordResponse[]
  total: number
}

// ===== 白名单 =====

export interface WhitelistCreate {
  term: string
  reason: string
  brand_id: string
}

export interface WhitelistResponse {
  id: string
  term: string
  reason: string
  brand_id: string
}

export interface WhitelistListResponse {
  items: WhitelistResponse[]
  total: number
}

// ===== 竞品 =====

export interface CompetitorCreate {
  name: string
  brand_id: string
  logo_url?: string
  keywords: string[]
}

export interface CompetitorResponse {
  id: string
  name: string
  brand_id: string
  logo_url?: string | null
  keywords: string[]
}

export interface CompetitorListResponse {
  items: CompetitorResponse[]
  total: number
}

// ===== 平台规则 =====

export interface PlatformRuleResponse {
  platform: string
  rules: Record<string, unknown>[]
  version: string
  updated_at: string
}

export interface PlatformListResponse {
  items: PlatformRuleResponse[]
  total: number
}

// ===== 品牌方平台规则（文档上传 + AI 解析） =====

export interface ParsedRulesData {
  forbidden_words: string[]
  restricted_words: { word: string; condition: string; suggestion: string }[]
  duration: { min_seconds?: number; max_seconds?: number } | null
  content_requirements: string[]
  other_rules: { rule: string; description: string }[]
}

export interface PlatformRuleParseRequest {
  document_url: string
  document_name: string
  platform: string
  brand_id: string
}

export interface PlatformRuleParseResponse {
  id: string
  platform: string
  brand_id: string
  document_url: string
  document_name: string
  parsed_rules: ParsedRulesData
  status: string
}

export interface PlatformRuleConfirmRequest {
  parsed_rules: ParsedRulesData
}

export interface BrandPlatformRuleResponse {
  id: string
  platform: string
  brand_id: string
  document_url: string
  document_name: string
  parsed_rules: ParsedRulesData
  status: string  // draft / active / inactive
  created_at: string
  updated_at: string
}

export interface BrandPlatformRuleListResponse {
  items: BrandPlatformRuleResponse[]
  total: number
}

// ===== 通用规则文档上传解析 =====

export interface RuleDocumentParseRequest {
  document_url: string
  document_name: string
  rule_type: 'forbidden_words' | 'whitelist' | 'competitors'
  brand_id?: string
}

export interface ParsedForbiddenWord {
  word: string
  category: string
  severity: string
}

export interface ParsedWhitelistItem {
  term: string
  reason: string
}

export interface ParsedCompetitor {
  name: string
  keywords: string[]
}

export interface RuleDocumentParseResponse {
  rule_type: string
  document_name: string
  forbidden_words: ParsedForbiddenWord[]
  whitelist_items: ParsedWhitelistItem[]
  competitors: ParsedCompetitor[]
  total_parsed: number
  duplicates_removed: number
}

export interface RuleDocumentConfirmRequest {
  rule_type: 'forbidden_words' | 'whitelist' | 'competitors'
  brand_id?: string
  forbidden_words: ParsedForbiddenWord[]
  whitelist_items: ParsedWhitelistItem[]
  competitors: ParsedCompetitor[]
}

export interface RuleDocumentConfirmResponse {
  added: number
  skipped_duplicates: number
  total: number
}

// ===== 品牌学习档案 =====

export interface LearnedRuleResponse {
  id: string
  type: string
  pattern: string
  reason: string
  source_task?: string | null
  created_by: string
  created_at?: string | null
}

export interface LearnedRuleCreateRequest {
  type: string
  pattern: string
  reason: string
}

// ===== 规则冲突检测 =====

export interface RuleValidateRequest {
  brand_id: string
  platform: string
  brief_rules: Record<string, unknown>
}

export interface RuleConflict {
  brief_rule: string
  platform_rule: string
  suggestion: string
}

export interface RuleValidateResponse {
  conflicts: RuleConflict[]
}

File: backend/app/schemas/rules.py
"""
平台规则相关 Schema
"""
from typing import Optional
from pydantic import BaseModel, Field


class PlatformRuleParseRequest(BaseModel):
    """上传文档并解析"""
    document_url: str = Field(..., description="TOS 上传后的文件 URL")
    document_name: str = Field(..., description="原始文件名（用于判断格式）")
    platform: str = Field(..., description="目标平台 (douyin/xiaohongshu/bilibili/kuaishou)")
    brand_id: str = Field("", description="品牌 ID（可选，后端自动从 tenant_id 获取）")


class ParsedRulesData(BaseModel):
    """AI 解析出的结构化规则"""
    forbidden_words: list[str] = Field(default_factory=list, description="违禁词列表")
    restricted_words: list[dict] = Field(
        default_factory=list,
        description="限制词 [{word, condition, suggestion}]",
    )
    duration: Optional[dict] = Field(
        None,
        description="时长要求 {min_seconds, max_seconds}",
    )
    content_requirements: list[str] = Field(
        default_factory=list,
        description="内容要求（如'必须展示产品'）",
    )
    other_rules: list[dict] = Field(
        default_factory=list,
        description="其他规则 [{rule, description}]",
    )


class PlatformRuleParseResponse(BaseModel):
    """解析响应（draft 状态）"""
    id: str
    platform: str
    brand_id: str
    document_url: str
    document_name: str
    parsed_rules: ParsedRulesData
    status: str


class PlatformRuleConfirmRequest(BaseModel):
    """确认/编辑解析结果"""
    parsed_rules: ParsedRulesData = Field(..., description="品牌方可能修改过的规则")


class PlatformRuleResponse(BaseModel):
    """完整响应"""
    id: str
    platform: str
    brand_id: str
    document_url: str
    document_name: str
    parsed_rules: ParsedRulesData
    status: str
    created_at: str
    updated_at: str


class PlatformRuleListResponse(BaseModel):
    """列表响应"""
    items: list[PlatformRuleResponse]
    total: int


# ==================== 通用规则文档解析 ====================

class RuleDocumentParseRequest(BaseModel):
    """通用规则文档解析请求"""
    document_url: str = Field(..., description="TOS 上传后的文件 URL")
    document_name: str = Field(..., description="原始文件名")
    rule_type: str = Field(..., description="规则类型: forbidden_words/whitelist/competitors")
    brand_id: str = Field("", description="品牌 ID（可选，后端自动从 tenant_id 获取）")


class ParsedForbiddenWord(BaseModel):
    word: str
    category: str = "custom"
    severity: str = "medium"


class ParsedWhitelistItem(BaseModel):
    term: str
    reason: str = ""


class ParsedCompetitor(BaseModel):
    name: str
    keywords: list[str] = Field(default_factory=list)


class RuleDocumentParseResponse(BaseModel):
    """通用规则文档解析响应（预览）"""
    rule_type: str
    document_name: str
    forbidden_words: list[ParsedForbiddenWord] = Field(default_factory=list)
    whitelist_items: list[ParsedWhitelistItem] = Field(default_factory=list)
    competitors: list[ParsedCompetitor] = Field(default_factory=list)
    total_parsed: int = 0
    duplicates_removed: int = 0


class RuleDocumentConfirmRequest(BaseModel):
    """确认批量写入"""
    rule_type: str = Field(..., description="规则类型: forbidden_words/whitelist/competitors")
    brand_id: str = Field("", description="品牌 ID（可选，后端自动从 tenant_id 获取）")
    forbidden_words: list[ParsedForbiddenWord] = Field(default_factory=list)
    whitelist_items: list[ParsedWhitelistItem] = Field(default_factory=list)
    competitors: list[ParsedCompetitor] = Field(default_factory=list)


class RuleDocumentConfirmResponse(BaseModel):
    """批量写入结果"""
    added: int = 0
    skipped_duplicates: int = 0
    total: int = 0
